keras_nlp/samplers/sampler.py

```diff
@@ -0,0 +1,279 @@
+        token_probability_fn,
+        prompt,
+        10,
+        end_token_id=tokenizer.token_to_id("[END]")
+    )
+    print(tokenizer.detokenize(generated))
+    ```
+    """
+
+    def __init__(
+        self,
+        jit_compile=True,
+    ):
+        self.jit_compile = jit_compile
+
+    def _validate_prompt_and_mask(self, prompt, mask):
+        """Helper method to validate input prompt."""
+        if not isinstance(prompt, (list, tf.RaggedTensor, tf.Tensor)):
+            raise ValueError(
+                "`prompt` must be one of `list`, `tf.RaggedTensor` or "
+                f"`tf.Tensor`, but received: prompt={type(prompt)}."
+            )
+
+        if isinstance(prompt, tf.RaggedTensor):
+            if mask:
+                raise ValueError(
+                    "`mask` is only valid when `prompt` is a list or dense "
+                    f"tensor, but received type(prompt)={type(prompt)}."
+                )
+            return prompt, mask
+
+        if isinstance(prompt, list):
+            prompt = tf.convert_to_tensor(prompt)
+        if not mask:
+            mask = tf.cast(tf.ones_like(prompt), dtype=tf.bool)
+        prompt = tf.ragged.boolean_mask(prompt, mask)
+        return prompt, mask
+
+    def _validate_token_probability_fn(
+        self, token_probability_fn, prompt, mask
+    ):
+        """Helper method to validate `token_probability_fn` output."""
+        test_pred = token_probability_fn(prompt, mask=mask)
+        if len(test_pred.shape) != 3:
+            raise ValueError(
+                "Output of `token_probability_fn` is not a 3D tensor, "
+                "please provide a function with the output shape "
+                "[batch_size, sequence_length, vocab_size]."
+            )
+
+    def _pad_prompt(self, prompt, max_length):
+        """Pad prompt to `max_length`."""
+        mask = tf.ones_like(prompt, dtype=tf.bool)
+        mask = mask.to_tensor(shape=(None, max_length))
+        prompt = prompt.to_tensor(shape=(None, max_length))
+        return prompt, mask
+
+    def _mask_tokens_after_end_token(
+        self,
+        prompt,
+        max_length,
+        end_token_id,
+    ):
+        """Helper function to truncate the tokens after the end token."""
+        # Mask out tokens after `end_token_id` is encountered.
+        # Find index of first end_token_id.
+        end_indices = tf.math.argmax(prompt == end_token_id, -1)
+        # Use max_length if no `end_token_id` is found.
+        end_indices = tf.where(
+            end_indices == 0,
+            tf.cast(max_length, dtype=end_indices.dtype),
+            end_indices,
+        )
+        # Truncate out tokens after (including) the end token.
+        mask_indices = tf.sequence_mask(end_indices, maxlen=max_length)
+        return tf.ragged.boolean_mask(prompt, mask_indices)
+
+    def __call__(
+        self,
+        token_probability_fn,
+        prompt,
+        max_length,
+        padding_mask=None,
```

Comment:
I would just call this `mask`. It is important we keep the same name for the input here and the arg we call `token_probability_fn` with. We want to make the link between the two clear!

Another option would be to call this `padding_mask` everywhere, but `mask` seems both more readable and technically correct? This is really just the mask of positions we will update. If it is not a padding mask, that is ok actually.